polynote-server/src/main/scala/polynote/server/repository/NotebookRepository.scala

```diff
@@ -155,6 +275,10 @@ class TreeRepository (
      } yield normalizePath(base.map(b => nb.copy(path = Paths.get(b, nb.path).toString)).getOrElse(nb))
   }
 
+  override def openNotebook(path: String): RIO[BaseEnv with GlobalEnv, NotebookRef] = delegate(path) {
+    (repo, relativePath, base) => TreeNotebookRef(repo, base, relativePath)
+  }
+
```

Comment:
One thing that's not ideal is that the repository itself doesn't stop you from opening multiple distinct `NotebookRef`s to the same notebook, which would overwrite each other's changes on disk. Currently the NotebookManager means that won't happen, but maybe the repository should also track open refs to avoid making duplicates. I guess we're kind of blurring the lines of NotebookManager/NotebookRepository, but I still think this separation is needed because repositories should have more control over how *changes* (not just whole notebooks) are persisted.

Reply:
do you think it's worth it to be paranoid / defensive here and prevent duplication?